### ISSUE TITLE
Develco Frient ZHEMI101 Energy Meter - Pulse Rate Adjust

### ISSUE THREAD

**Issue by lanky8804** [Koenkk/zigbee2mqtt#8638]
With the new addition of the energy monitoring dashboard, how can I set the pulse rate of my electric meter for the Frient zhemi101?

My pulse rate is 3200 imp/kWh

**Comment by robertalexa** [comment]
@lanky8804 This is a duplicate of https://github.com/Koenkk/zigbee2mqtt/discussions/8405

Can you please close this issue and let's talk there. I don't have the device but I can try to help you guys out.

**Comment by robertalexa** [comment]
![image](https://user-images.githubusercontent.com/9019306/132094764-1ef90fef-adca-47ee-8863-ec239f61594a.png)

Needed attribute

**Comment by robertalexa** [comment]
Create PR to zigbee-herdsman to add support for the attribute:
https://github.com/Koenkk/zigbee-herdsman/pull/413

I will look at commands soon and get you to help with testing please. As mentioned before, i do not own the device so I am working blindly here.

**Comment by grudyjsy** [comment]
> Create PR to zigbee-herdsman to add support for the attribute:
> [Koenkk/zigbee-herdsman#413](https://github.com/Koenkk/zigbee-herdsman/pull/413)
> 
> I will look at commands soon and get you to help with testing please. As mentioned before, i do not own the device so I am working blindly here.

Thank you, happy to test when ever you need me to.

**Comment by robertalexa** [comment]
Will continue as soon as time allows. I have a few bits started already that i need to finish before, but I will keep you updated.

**Comment by lanky8804** [comment]
Thank you, also happy to test anything

**Comment by grudyjsy** [comment]
Just in case you want to have some sensors that are more accurate in the meantime I have created these:

template:
  - sensor:
      - name: "Electricity Live Usage"
        unit_of_measurement: W
        state_class: measurement
        device_class: power
        state: "{{ (states('sensor.develco_energy_meter_power')|float * 1000 / 3200) | round(2) }}"

      - name: "Electricity Usage Total"
        unit_of_measurement: kWh
        state_class: measurement
        device_class: energy
        state: "{{ (states('sensor.develco_energy_meter_energy')|float * 1000 / 3200) | round(2) }}"


I can't seem to get them to work in the energy module but the are ok for a dashboard.

**Comment by pascalschwientek** [comment]
Hello! Just got this device and I need to set it to 10000 per kwh.
I saw that the PR from @robertalexa in zigbee-herdsman is merged. I installed the edge version of zigbee2mqtt, as the dev branch already includes the updated zigbee-herdsman library.

I am still new to Home Assistant, ZigBee and zigbee2mqtt. But I would expect 'develcoPulseConfiguration' to show up in the dev console for the device under 'seMetering'. But it did not show up. So I am guessing this new config still needs to be mentioned in the zigbee2mqtt codebase. I tried looking through the codebase, but didn't see anything obvious where this would make sense.

@robertalexa thanks for the help on this even though you don't even own the device. Do you maybe have the time to point me in the right direction so that I could try my luck on it?

**Comment by robertalexa** [comment]
Hey guys, sorry I have been absent, I have not forgot about this, life has just kept me busier than anticipated.

You will basically need to try this with an external converter. In your converter you will will to provide the functions to handle those commands, just like fz. and tz. do, only that you first try it as functions in your converter.

One good example of custom commands can be found for the Danfoss Ally thermostat (that is the first device that comes to mind). Have a look at that device, see if you can figure it out. I will try to find some time ASAP to provide an external converter for you, in case you don't figure it out before me.

**Comment by pascalschwientek** [comment]
Thanks! I tried, and I think I am on the right track. Either the data regarding the pulse configuration is not there or I am not actually on the right track.

```javascript
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const constants = require('zigbee-herdsman-converters/lib/constants');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;


const options = {manufacturerCode: 0x1015};

dpcTZ = {
    key: ['pulse_configuration'],
    convertSet: async (entity, key, value, meta) => {
        await entity.write('seMetering', {'develcoPulseConfiguration': value}, options);
        return {readAfterWriteTime: 200, state: {'thermostat_vertical_orientation': value}};
    },
    convertGet: async (entity, key, meta) => {
        await entity.read('seMetering', ['develcoPulseConfiguration'], options);
    },
};

const definition = {
    zigbeeModel: ['ZHEMI101'],
    model: 'ZHEMI101',
    vendor: 'Develco',
        description: 'Energy meter',
    fromZigbee: [fz.metering, dpcTZ],
    toZigbee: [dpcTZ],
    configure: async (device, coordinatorEndpoint, logger) => {
        const endpoint = device.getEndpoint(2);

        await reporting.bind(endpoint, coordinatorEndpoint, ['seMetering']);
        await reporting.instantaneousDemand(endpoint);
        await reporting.readMeteringMultiplierDivisor(endpoint);
    },
    exposes: [e.power(), e.energy(),
        exposes.numeric('pulse_configuration', ea.ALL)
        .withDescription('Pulses per kwh')
    ],
};

module.exports = definition;
```

Log message:
`Error Publish 'get' 'pulse_configuration' to 'Energy meter' failed: 'Error: Read 0x0015bc001b021f57/1 seMetering(["develcoPulseConfiguration"], {"sendWhenActive":false,"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":4117,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Timeout - 41423 - 1 - 5 - 1794 - 1 after 10000ms)'`


I have also seen this debug message, which makes me believe that there is no pulse in the data yet.

`Received Zigbee message from 'Energy meter', type 'attributeReport', cluster 'seMetering', data '{"currentSummDelivered":[0,85789],"currentTier1SummDelivered":[65535,4294967295],"currentTier2SummDelivered":[65535,4294967295],"instantaneousDemand":3798,"status":0}' from endpoint 2 with groupID 0`

**Comment by lanky8804** [comment]
> Just in case you want to have some sensors that are more accurate in the meantime I have created these:
> 
> template:
> 
> * sensor:
>   
>   * name: "Electricity Live Usage"
>     unit_of_measurement: W
>     state_class: measurement
>     device_class: power
>     state: "{{ (states('sensor.develco_energy_meter_power')|float * 1000 / 3200) | round(2) }}"
>   * name: "Electricity Usage Total"
>     unit_of_measurement: kWh
>     state_class: measurement
>     device_class: energy
>     state: "{{ (states('sensor.develco_energy_meter_energy')|float * 1000 / 3200) | round(2) }}"
> 
> I can't seem to get them to work in the energy module but they are ok for a dashboard.

Thank you this is a very nice temp fix

**Comment by robertalexa** [comment]
@pascalschwientek Hey, i just tried to put sth together, but I have no way of testing. Try to use this code and tell me what happens:

```javascript
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;

const fzLocal = {
    develco_metering: {
        cluster: 'seMetering',
        type: ['attributeReport', 'readResponse'],
        convert: (model, msg, publish, options, meta) => {
            const result = {};
            if (msg.data.hasOwnProperty('develcoPulseConfiguration')) {
                result[postfixWithEndpointName('pulse_configuration', msg, model)] =
                    msg.data['develcoPulseConfiguration'];
            }
            return result;
        },
    },
}

const tzLocal = {
    develco_pulse_configuration: {
        key: ['pulse_configuration'],
        convertSet: async (entity, key, value, meta) => {
            await entity.write('seMetering', {'develcoPulseConfiguration': value}, 0x1015);
            return {readAfterWriteTime: 200, state: {'pulse_configuration': value}};
        },
        convertGet: async (entity, key, meta) => {
            await entity.read('seMetering', ['develcoPulseConfiguration'], 0x1015);
        },
    },
}

const definition = {
    zigbeeModel: ['ZHEMI101'],
    model: 'ZHEMI101',
    vendor: 'Develco',
    description: 'Energy meter',
    fromZigbee: [fz.metering, fzLocal.develco_metering],
    toZigbee: [tzLocal.develco_pulse_configuration],
    configure: async (device, coordinatorEndpoint, logger) => {
        const endpoint = device.getEndpoint(2);
        await reporting.bind(endpoint, coordinatorEndpoint, ['seMetering']);
        await reporting.instantaneousDemand(endpoint);
        await reporting.readMeteringMultiplierDivisor(endpoint);
    },
    exposes: [e.power(), e.energy(),
    exposes.numeric('pulse_configuration', ea.ALL)
        .withDescription('Pulses per kwh')],
};

module.exports = definition;
```

**Comment by pascalschwientek** [comment]
Same Error message:

`Received MQTT message on 'zigbee2mqtt/Energy meter/set' with data '{"pulse_configuration":10000}'`

`Publish 'set' 'pulse_configuration' to 'Energy meter' failed: 'Error: Write 0x0015bc001b021f57/1 seMetering({"develcoPulseConfiguration":10000}, {"sendWhenActive":false,"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Timeout - 41423 - 1 - 124 - 1794 - 4 after 10000ms)'`

I am running version 1.21.1-dev commit: eca5492
which has the develcoPulseConfiguration definition included.

**Comment by robertalexa** [comment]
my bad with the manufacturer code!

Can you replace `0x1015` with `{manufacturerCode: 0x1015}` on both lines that use it

![image](https://user-images.githubusercontent.com/9019306/133426097-ecdf97a8-ba9c-49cf-8060-71c616dc541e.png)

Let me know how it goes

**Comment by pascalschwientek** [comment]
No success, same error messages. 

I am wondering, this is the data payload we get from the device according to the logs:
`{"currentSummDelivered":[0,103522],"currentTier1SummDelivered":[65535,4294967295],"currentTier2SummDelivered":[65535,4294967295],"instantaneousDemand":6930,"status":0}`

shouldn't the object contain develcoPulseConfiguration?

**Comment by pascalschwientek** [comment]
Btw, the device got re-branded. It's not called "frient". Might this have also caused a new manufacturer code? But then it wouldn't make sense that we get all other data perfectly fine from the device.

https://zigbee.blakadder.com/Develco_ZHEMI101.html

**Comment by robertalexa** [comment]
Just to double check sth @pascalschwientek 

You are setting this up using an external converter? If so, in your database.db file, edit your device model number. For example add `-test` to it. Then in the external converter change this line `model: 'ZHEMI101'` to the new name.

This is to make sure that only the external converter code is running and not the current implementation.

The values you are seeing are most likely coming from the device Reporting. For the pulse_configuration attribute there would not be any reporting, just manual GET and SET (when needed, probably just once in a household).

Once you do the above, can you run a GET command and also a SET command and post the error message please (if you get one.) I want to make sure the correct manufacturer is passed through etc.

**Comment by robertalexa** [comment]
@pascalschwientek This has gotten me puzzled. I can't spot anything wrong in what we did.... From the other people following this thread, is anyone else able to test this using the dev branch? (so that the latest zigbe--herdsman is used)

@sjorge or @Koenkk Can you guys have a glance at this and tell me if I am missing something obvious? Thanks!

**Comment by sjorge** [comment]
Timeout seems to indicate the device is not responding at all I'm guessing.

**Comment by robertalexa** [comment]
Those were my thoughts exactly, but @pascalschwientek seems to be getting data back from the device about the power consumption, which made me even more confused.

**Comment by pascalschwientek** [comment]
```
Info MQTT publish: topic 'zigbee2mqtt/Energy meter', payload '{"energy":116.43,"linkquality":132,"power":6669,"pulse_configuration":null}'
Info MQTT publish: topic 'zigbee2mqtt/Energy meter', payload '{"energy":116.44,"linkquality":132,"power":6974,"pulse_configuration":null}'
Info MQTT publish: topic 'zigbee2mqtt/Energy meter', payload '{"energy":116.45,"linkquality":132,"power":6925,"pulse_configuration":null}'
Info MQTT publish: topic 'zigbee2mqtt/Energy meter', payload '{"energy":116.46,"linkquality":33,"power":7102,"pulse_configuration":null}'
Info MQTT publish: topic 'zigbee2mqtt/Energy meter', payload '{"energy":116.47,"linkquality":39,"power":7206,"pulse_configuration":null}'
Info MQTT publish: topic 'zigbee2mqtt/Energy meter', payload '{"energy":116.48,"linkquality":36,"power":6956,"pulse_configuration":null}'
Error Publish 'set' 'pulse_configuration' to 'Energy meter' failed: 'Error: Write 0x0015bc001b021f57/1 seMetering({"develcoPulseConfiguration":19238}, {"sendWhenActive":false,"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":4117,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Timeout - 41423 - 1 - 16 - 1794 - 4 after 10000ms)'
Info MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"message":"Publish 'set' 'pulse_configuration' to 'Energy meter' failed: 'Error: Write 0x0015bc001b021f57/1 seMetering({\"develcoPulseConfiguration\":19238}, {\"sendWhenActive\":false,\"timeout\":10000,\"disableResponse\":false,\"disableRecovery\":false,\"disableDefaultResponse\":true,\"direction\":0,\"srcEndpoint\":null,\"reservedBits\":0,\"manufacturerCode\":4117,\"transactionSequenceNumber\":null,\"writeUndiv\":false}) failed (Timeout - 41423 - 1 - 16 - 1794 - 4 after 10000ms)'","meta":{"friendly_name":"Energy meter"},"type":"zigbee_publish_error"}'
Info MQTT publish: topic 'zigbee2mqtt/Energy meter', payload '{"energy":116.5,"linkquality":36,"power":7061,"pulse_configuration":null}'
Info MQTT publish: topic 'zigbee2mqtt/Energy meter', payload '{"energy":116.51,"linkquality":33,"power":6887,"pulse_configuration":null}'
Info MQTT publish: topic 'zigbee2mqtt/Energy meter', payload '{"energy":116.52,"linkquality":33,"power":7625,"pulse_configuration":null}'
```

**Comment by sjorge** [comment]
Might be that coordinator -> device has a different route than device -> coordinator.

**Comment by pascalschwientek** [comment]
How could I check if this might be a problem?

**Comment by sjorge** [comment]
You could try power cycling a few nearby router devices and hope it picks a new route, although Koen might have a better answer

**Comment by robertalexa** [comment]
@grudyjsy @lanky8804 Can you guys try to test this as well and see what results you are getting? Thanks

**Comment by lanky8804** [comment]
@robertalexa If someone can give me instructions on how to do this then can test.

**Comment by robertalexa** [comment]
@lanky8804 You will need to use the dev version of z2m, add an external converter (instruction in the documentation) and use the code provided above

**Comment by pascalschwientek** [comment]
@lanky8804 how is your setup looking? If you are running HomeAssistant OS with the Hassio Add-On:

Backup config from: Supervisor → Dashboard → Zigbee2mqtt → Configuration
(Uninstall if already installed: Supervisor → Dashboard → Zigbee2mqtt Edge → Uninstall)
Refresh repo: Supervisor → Add-on store → ⋮ → Reload
Re-install: Supervisor → Add-on store → Zigbee2mqtt Edge → Install
Restore config to: Supervisor → Dashboard → Zigbee2mqtt Edge → Configuration

Then inside the zigbee2mqtt folder (where the configuration.yaml is) create a new file called: `ZHEMI101.js ` with the following content:

```javascript
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;

const fzLocal = {
    develco_metering: {
        cluster: 'seMetering',
        type: ['attributeReport', 'readResponse'],
        convert: (model, msg, publish, options, meta) => {
            const result = {};
            if (msg.data.hasOwnProperty('develcoPulseConfiguration')) {
                result[postfixWithEndpointName('pulse_configuration', msg, model)] =
                    msg.data['develcoPulseConfiguration'];
            }
            return result;
        },
    },
}

const tzLocal = {
    develco_pulse_configuration: {
        key: ['pulse_configuration'],
        convertSet: async (entity, key, value, meta) => {
            await entity.write('seMetering', {'develcoPulseConfiguration': value}, {manufacturerCode: 0x1015});
            return {readAfterWriteTime: 200, state: {'pulse_configuration': value}};
        },
        convertGet: async (entity, key, meta) => {
            await entity.read('seMetering', ['develcoPulseConfiguration'], {manufacturerCode: 0x1015});
        },
    },
}

const definition = {
    zigbeeModel: ['ZHEMI101'],
    model: 'ZHEMI101',
    vendor: 'Develco',
    description: 'Energy meter',
    fromZigbee: [fz.metering, fzLocal.develco_metering],
    toZigbee: [tzLocal.develco_pulse_configuration],
    configure: async (device, coordinatorEndpoint, logger) => {
        const endpoint = device.getEndpoint(2);
        await reporting.bind(endpoint, coordinatorEndpoint, ['seMetering']);
        await reporting.instantaneousDemand(endpoint);
        await reporting.readMeteringMultiplierDivisor(endpoint);
    },
    exposes: [
        e.power(), 
        e.energy(),
        exposes.numeric('pulse_configuration', ea.ALL).withValueMin(0).withValueMax(65535)
            .withDescription('Pulses per kwh. Default 1000 imp/kWh')],
};

module.exports = definition;
```

Now add it to the configuration. Either edit the file or go back to Supervisor → Dashboard → Zigbee2mqtt Edge → Configuration 

```
external_converters:
  - ZHEMI101.js
```

now restart and that should be it.

**Comment by lanky8804** [comment]
@pascalschwientek Running HA on a RasPi 4. Just installed the Edge version which is now up and running. Will now follow other instructions (thank you) and get back to everyone

**Comment by lanky8804** [comment]
@pascalschwientek @robertalexa ok so done all above, thank you.

Getting this in the log:
`Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":974.16,"linkquality":26,"power":2431,"pulse_configuration":null}'
Error Publish 'get' 'pulse_configuration' to 'Frient - Energy Monitor' failed: 'Error: Read 0x0015bc001b02246d/1 seMetering(["develcoPulseConfiguration"], {"sendWhenActive":false,"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":4117,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Timeout - 7240 - 1 - 4 - 1794 - 1 after 10000ms)'
Info MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"message":"Publish 'get' 'pulse_configuration' to 'Frient - Energy Monitor' failed: 'Error: Read 0x0015bc001b02246d/1 seMetering([\"develcoPulseConfiguration\"], {\"sendWhenActive\":false,\"timeout\":10000,\"disableResponse\":false,\"disableRecovery\":false,\"disableDefaultResponse\":true,\"direction\":0,\"srcEndpoint\":null,\"reservedBits\":0,\"manufacturerCode\":4117,\"transactionSequenceNumber\":null,\"writeUndiv\":false}) failed (Timeout - 7240 - 1 - 4 - 1794 - 1 after 10000ms)'","meta":{"friendly_name":"Frient - Energy Monitor"},"type":"zigbee_publish_error"}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":974.17,"linkquality":26,"power":2423,"pulse_configuration":null}'`

Also when I try and change pulse rate in the device exposes it doesnt save.

**Comment by robertalexa** [comment]
@sjorge Same error as @pascalschwientek. So we know it is not connection.

**Comment by sjorge** [comment]
Have you tried sending it without the manufacturerCode ? Some ubisys devices refuse some reds too if we specify it... hence there is a null variant too.

**Comment by robertalexa** [comment]
Documentation specifically mentions manufacturer code:
![image](https://user-images.githubusercontent.com/9019306/133472193-f30f98ca-2f48-41ee-87ed-4dc4cd68c41c.png)

By mistake we did run some commands with manufacturer code null and it still failed: 

> Same Error message:
> 
> `Received MQTT message on 'zigbee2mqtt/Energy meter/set' with data '{"pulse_configuration":10000}'`
> 
> `Publish 'set' 'pulse_configuration' to 'Energy meter' failed: 'Error: Write 0x0015bc001b021f57/1 seMetering({"develcoPulseConfiguration":10000}, {"sendWhenActive":false,"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Timeout - 41423 - 1 - 124 - 1794 - 4 after 10000ms)'`
> 
> I am running version 1.21.1-dev commit: [eca5492](https://github.com/Koenkk/zigbee2mqtt/commit/eca5492a8e9eff9d521426eefe23e4800e080c39)
> which has the develcoPulseConfiguration definition included.

**Comment by lanky8804** [comment]
@robertalexa Some of this goes beyond my understanding but could we test a hardcoded pulsecounter value to see if that worked? Rather than the variable one?

How do we find out the version the device is running on? Just thinking as this is rebranded as mentioned before might be an older version??

**Comment by sjorge** [comment]
I did notice it is marked as Optional, maybe Friend didn't implement it?

**Comment by robertalexa** [comment]
> @robertalexa Some of this goes beyond my understanding but could we test a hardcoded pulsecounter value to see if that worked? Rather than the variable one?

That is what we are trying to do. The value needs to be SET to the device, it does not live in z2m. But as you confirmed, the command fails.



> I did notice it is marked as Optional, maybe Friend didn't implement it?

The documentation manual is for the device, I think Friend just rebranded, but don;t know. Also don't own it :)

**Comment by pascalschwientek** [comment]
In the docs they mention SW versions. (I guess software version?)
Is there a way to read the sw version of the hardware?

`The Basic clustercontains a MFG attributewith theSW version.MFG code: 0x1015 [Develco], Attribute id: 0x8000, Value [3 Byte]Example: 0x040601is SW4.6.1`

**Comment by robertalexa** [comment]
> Is there a way to read the sw version of the hardware?

Only if you can edit a file on your zigbee-herdsman to include that attribute. That is also manufacturer specific and is currently not present in zigbee-herdsman.

You would needs to edit the cluster file and under `genBasic` you would need to add a command for the attribute, in a similar way to how I added the `develcoPulseConfiguration` attribute

**Comment by robertalexa** [comment]
What I am finding strange is that the pulse_configuration attribute was successfully added to deconz and a lot of people already confirmed it working, so we are either doing sth wrong, or we are indeed unlucky and have problems with the device SW
https://github.com/dresden-elektronik/deconz-rest-plugin/pull/4426/commits/3a63eb6799e45b1f11fcdcd25839af4af6c6254d

**Comment by pascalschwientek** [comment]
Will try to get the SW version later tonight when I am home.

Very strange. I have asked frient for some technical documentation. I hope they reply with something useful.

**Comment by sjorge** [comment]
They are setting both in one payload if I read the code correctly?
```
                   sensor.addItem(DataTypeUInt8, RConfigInterfaceMode)->setValue(1);
                    sensor.addItem(DataTypeUInt16, RConfigPulseConfiguration)->setValue(1000);
```

Maybe they both need to be set or CfgIntfMode needs to be set first?

**Comment by robertalexa** [comment]
I have just done a new PR to zigbee-herdsman for this other attribute, maybe they are indeed paired.

**Comment by robertalexa** [comment]
![image](https://user-images.githubusercontent.com/9019306/133478391-eefe718f-0855-4ea7-be3c-e82fbd8471a7.png)

It is also weird that Deconz are setting the value to 1 for interface mode, which would be `Pulse probe –Pulse counting on a gas meter`. Something doesn't add up.

**Comment by pascalschwientek** [comment]
From the deconz PR that introduced pulse configuration:
`Also the interface mode had some blemishes... Btw, allowed interface modes are from 1 (should be default, in order according to documentation, to 8).`

**Comment by robertalexa** [comment]
Yeah, it isn't actually 1 to 8, they are building a list and assigning values. These are the values as per docs:
![image](https://user-images.githubusercontent.com/9019306/133485968-14090d34-79ae-4f6e-a4ac-d9aea719b74b.png)

**Comment by pascalschwientek** [comment]
Because zigbee2mqtt is in docker, I didn't find a quick way to edit the files. I probably need to edit the docker file first to mount the app folder and then edit from the host machine.

But I also updated to the latest edge version today and tried the interface mode - same story:

`Publish 'get' 'interface_mode' to 'Energy meter' failed: 'Error: Read 0x0015bc001b021f57/1 seMetering(["develcoInterfaceMode"], {"sendWhenActive":false,"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":4117,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Timeout - 41423 - 1 - 49 - 1794 - 1 after 10000ms)'`

`Publish 'set' 'interface_mode' to 'Energy meter' failed: 'Error: Write 0x0015bc001b021f57/1 seMetering({"develcoInterfaceMode":0}, {"sendWhenActive":false,"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":4117,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Timeout - 41423 - 1 - 123 - 1794 - 4 after 10000ms)'`

I wanted to buy a product to just have it work plug and play - should have gone the DIY route with ESPHome 😂

**Comment by robertalexa** [comment]
> I wanted to buy a product to just have it work plug and play - should have gone the DIY route with ESPHome

I can understand the frustration. Hopefully we will get to the bottom of it. I have asked Koen to have a look at this and see if he has any thoughts. I think in advance, can you provide the zigbee-herdsman log too when running the commands?

**Comment by lanky8804** [comment]
> I wanted to buy a product to just have it work plug and play - should have gone the DIY route with ESPHome 😂

I also looked into this route befor purchasing but powering the device was going to be the issue for me, hense this all in one.

Appriciate the support, hopefully we can resolve the issue

**Comment by pascalschwientek** [comment]
[logs.txt](https://github.com/Koenkk/zigbee2mqtt/files/7176930/logs.txt)

I don't know what to look for - so I didn't filter out anything, just the raw data. I did a set for both and a get for both. I hope there isn't too much spam in the log from my other devices.


The Converter:

```javascript
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;

const fzLocal = {
    develco_metering: {
        cluster: 'seMetering',
        type: ['attributeReport', 'readResponse'],
        convert: (model, msg, publish, options, meta) => {
            const result = {};
            if (msg.data.hasOwnProperty('develcoPulseConfiguration')) {
                result[postfixWithEndpointName('pulse_configuration', msg, model)] =
                    msg.data['develcoPulseConfiguration'];
            }
            if (msg.data.hasOwnProperty('develcoInterfaceMode')) {
                result[postfixWithEndpointName('interface_mode', msg, model)] =
                    msg.data['develcoInterfaceMode'];
            }
            return result;
        },
    },
}

const tzLocal = {
    develco_pulse_configuration: {
        key: ['pulse_configuration'],
        convertSet: async (entity, key, value, meta) => {
            await entity.write('seMetering', {'develcoPulseConfiguration': value}, {manufacturerCode: 0x1015});
            return {readAfterWriteTime: 200, state: {'pulse_configuration': value}};
        },
        convertGet: async (entity, key, meta) => {
            await entity.read('seMetering', ['develcoPulseConfiguration'], {manufacturerCode: 0x1015});
        },
    },
    develco_interface_mode: {
        key: ['interface_mode'],
        convertSet: async (entity, key, value, meta) => {
            await entity.write('seMetering', {'develcoInterfaceMode': value}, {manufacturerCode: 0x1015});
            return {readAfterWriteTime: 200, state: {'interface_mode': value}};
        },
        convertGet: async (entity, key, meta) => {
            await entity.read('seMetering', ['develcoInterfaceMode'], {manufacturerCode: 0x1015});
        },
    },
}

const definition = {
    zigbeeModel: ['ZHEMI101'],
    model: 'ZHEMI101',
    vendor: 'Develco',
    description: 'Energy meter',
    fromZigbee: [fz.metering, fzLocal.develco_metering],
    toZigbee: [tzLocal.develco_pulse_configuration, tzLocal.develco_interface_mode],
    configure: async (device, coordinatorEndpoint, logger) => {
        const endpoint = device.getEndpoint(2);
        await reporting.bind(endpoint, coordinatorEndpoint, ['seMetering']);
        await reporting.instantaneousDemand(endpoint);
        await reporting.readMeteringMultiplierDivisor(endpoint);
    },
    exposes: [
        e.power(), 
        e.energy(),
        exposes.numeric('pulse_configuration', ea.ALL).withValueMin(0).withValueMax(65535)
            .withDescription('Pulses per kwh. Default 1000 imp/kWh'),
        exposes.numeric('interface_mode', ea.ALL).withValueMin(0).withValueMax(65535)
    ],
};

module.exports = definition;
```

**Comment by pascalschwientek** [comment]
Has anyone got a new idea on this? Did Koen have time to look at this already? Sadly, I am not getting any response from Frient regarding a current technical documentation.

**Comment by grudyjsy** [comment]
Hi, I have tried Frient via social media and email and heard nothing back. I wonder if Develco would be more help as it’s there products that Frient label up as thee own product, https://www.develcoproducts.com/products/meter-interfaces/external-meter-interface/

**Comment by robertalexa** [comment]
The documentation we have been following is Develco's
https://www.develcoproducts.com/media/1782/zhemi101-technical-manual-external-meter-interface.pdf

@Koenkk Would you be kind to have a look and see if we are missing anything here? Thanks in advance

**Comment by Koenkk** [comment]
Can you try with `await entity.write('seMetering', {'develcoInterfaceMode': value}, {manufacturerCode: 0x1015, sendWhenActive: true});` instead? This should send the write command once the device is awake (e.g. trigger the device to send a message to z2m)

**Comment by pascalschwientek** [comment]
Tried it but sadly no change in behavior. Since Frient is not responding, maybe it's worth a shot to purchase the develco product and rule out any firmware change that Frient made that's specific to them.

`Publish 'get' 'pulse_configuration' to 'Energy meter' failed: 'Error: Read 0x0015bc001b021f57/1 seMetering(["develcoPulseConfiguration"], {"sendWhenActive":false,"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":4117,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Timeout - 41423 - 1 - 29 - 1794 - 1 after 10000ms)'
`
`Publish 'set' 'pulse_configuration' to 'Energy meter' failed: 'Error: Write 0x0015bc001b021f57/1 seMetering({"develcoPulseConfiguration":0}, {"sendWhenActive":true,"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":4117,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Timeout - 41423 - 1 - 131 - 1794 - 4 after 10000ms)'`

**Comment by pascalschwientek** [comment]
I got a reply from Develco:

_We are the owner of the Frient brand._

I would rule out any firmware change. Develco is the B2B whitelabel manufacturer and Frient is their consumer brand.
But they also won't sell a Develco product to me as it's only B2B

**Comment by zuzuman** [comment]
Latest Zigbee2Mqtt:
![image](https://user-images.githubusercontent.com/3532812/140977057-3e94b93a-1de2-4ba7-8714-7c2ff39aaed9.png)
Can read values, but when trying to change them:
![image](https://user-images.githubusercontent.com/3532812/140977348-d7501ae8-db0d-4fb5-8442-304d9ae5ee95.png)

**Comment by zuzuman** [comment]
OK. Done this one

**Comment by zuzuman** [comment]
`const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;

const fzLocal = {
    develco_metering: {
        cluster: 'seMetering',
        type: ['attributeReport', 'readResponse'],
        convert: (model, msg, publish, options, meta) => {
            const result = {};
            if (msg.data.hasOwnProperty('develcoPulseConfiguration')) {
                result[postfixWithEndpointName('Pulse Configuration', msg, model)] =
                    msg.data['develcoPulseConfiguration'];
            }
            return result;
        },
    },
}

const tzLocal = {
    develco_interface_mode: {
        key: ['Pulse Configuration'],
        convertSet: async (entity, key, value, meta) => {

            await entity.write('seMetering', {'develcoPulseConfiguration' : value}, {manufacturerCode: 0x1015});
            return {readAfterWriteTime: 200, state: {'develcoPulseConfiguration': value}};
        },
        convertGet: async (entity, key, meta) => {
            await entity.read('seMetering', ['develcoPulseConfiguration'], {manufacturerCode: 0x1015});
        },
    },
}

const definition = {
    zigbeeModel: ['ZHEMI101'],
    model: 'ZHEMI101',
    vendor: 'Develco',
    description: 'Energy meter',
    fromZigbee: [fz.metering, fzLocal.develco_metering],
    endpoint: (device) => {
    return {'default': 2};
    },
    toZigbee: [tzLocal.develco_interface_mode],
    configure: async (device, coordinatorEndpoint, logger) => {
        const endpoint = device.getEndpoint(2);
        await reporting.bind(endpoint, coordinatorEndpoint, ['seMetering']);
        await reporting.instantaneousDemand(endpoint);
        await reporting.readMeteringMultiplierDivisor(endpoint);
    },
    exposes: [
        e.power(), 
        e.energy(),
        exposes.numeric('Pulse Configuration', ea.ALL).withDescription('Pulses per whole value (m³ or kWh)')
    ],
};

module.exports = definition;
`

Tested with electricity meter (2500 led impulses )
Tested with gas meter (replaced LED with magnetic sensor) (100 magnetic impulses per 1m³)

Do not use "refresh" button. Just provide value inside text box

**Comment by zuzuman** [comment]
https://user-images.githubusercontent.com/3532812/141027383-4fa6b10b-934f-4841-927f-af4375fe776c.mov

**Comment by lanky8804** [comment]
@zuzuman just tired this also with Edge version and 3200 pulse and seem to be working, thank you.

**Comment by robertalexa** [comment]
@zuzuman You have mixed names of attributes everywhere :) `develco_interface_mode` but in fact you use `develcoPulseConfiguration`

Both Interface mode and pulse configuration should be exposed as 2 separate attributes. They both accept read and write as well.

The reason why both attributes are needed is because interface mode needs to be changed accordingly if you change from electricity meter to gas meter, as seen in the documentation. See post https://github.com/Koenkk/zigbee2mqtt/issues/8638#issuecomment-920209163

Keep the key of the attribute as `pulse_configuration` and not `Pulse Configuration`. It is a key (not a name) that will be used as mqtt messages in json format

Would you please be kind to test with the original code, to which I have added the default endpoint now, and let me know how it goes?

```javascript
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;

const fzLocal = {
    develco_metering: {
        cluster: 'seMetering',
        type: ['attributeReport', 'readResponse'],
        convert: (model, msg, publish, options, meta) => {
            const result = {};
            if (msg.data.hasOwnProperty('develcoPulseConfiguration')) {
                result[postfixWithEndpointName('pulse_configuration', msg, model)] =
                    msg.data['develcoPulseConfiguration'];
            }
            if (msg.data.hasOwnProperty('develcoInterfaceMode')) {
                result[postfixWithEndpointName('interface_mode', msg, model)] =
                    msg.data['develcoInterfaceMode'];
            }
            return result;
        },
    },
}

const tzLocal = {
    develco_pulse_configuration: {
        key: ['pulse_configuration'],
        convertSet: async (entity, key, value, meta) => {
            await entity.write('seMetering', {'develcoPulseConfiguration': value}, {manufacturerCode: 0x1015});
            return {readAfterWriteTime: 200, state: {'pulse_configuration': value}};
        },
        convertGet: async (entity, key, meta) => {
            await entity.read('seMetering', ['develcoPulseConfiguration'], {manufacturerCode: 0x1015});
        },
    },
    develco_interface_mode: {
        key: ['interface_mode'],
        convertSet: async (entity, key, value, meta) => {
            await entity.write('seMetering', {'develcoInterfaceMode': value}, {manufacturerCode: 0x1015});
            return {readAfterWriteTime: 200, state: {'interface_mode': value}};
        },
        convertGet: async (entity, key, meta) => {
            await entity.read('seMetering', ['develcoInterfaceMode'], {manufacturerCode: 0x1015});
        },
    },
}

const definition = {
    zigbeeModel: ['ZHEMI101'],
    model: 'ZHEMI101',
    vendor: 'Develco',
    description: 'Energy meter',
    fromZigbee: [fz.metering, fzLocal.develco_metering],
    toZigbee: [tzLocal.develco_pulse_configuration, tzLocal.develco_interface_mode],
    endpoint: (device) => {
        return {'default': 2};
    },
    configure: async (device, coordinatorEndpoint, logger) => {
        const endpoint = device.getEndpoint(2);
        await reporting.bind(endpoint, coordinatorEndpoint, ['seMetering']);
        await reporting.instantaneousDemand(endpoint);
        await reporting.readMeteringMultiplierDivisor(endpoint);
    },
    exposes: [
        e.power(), 
        e.energy(),
        exposes.numeric('pulse_configuration', ea.ALL).withValueMin(0).withValueMax(65535)
            .withDescription('Pulses per kwh. Default 1000 imp/kWh'),
        exposes.numeric('interface_mode', ea.ALL).withValueMin(0).withValueMax(65535)
    ],
};

module.exports = definition;
```

@pascalschwientek Can you also give it a try?

**Comment by zuzuman** [comment]
@robertalexa 
This device can store any string/value pair in memory, and "develcoInterfaceMode" is just a configuration option, and it's not utilised by zigbee2mqtt, and don't know if will ever be (because each person creating own sensor with unity value). 

CurrentSummation - if you set it, metering will reset to 0. And again - you need to calculate by your own meter total value. 

.withValueMin(0).withValueMax(65535) - sliders are not working correctly in zigbee2Mqtt. It's sending 2-3 requests, and only first one contains selected value. 

msg.data.hasOwnProperty - is not populating selected value, because there are no property in message.

**Comment by lanky8804** [comment]
@robertalexa Have just updated to the above code and seems to be working with the device updating. This is Elec only for me.

**Comment by robertalexa** [comment]
@zuzuman Have you actually tried the code or are you just giving me your thoughts?

> This device can store any string/value pair in memory, and "develcoInterfaceMode" is just a configuration option, and it's not utilised by zigbee2mqtt, and don't know if will ever be (because each person creating own sensor with unity value).

This makes absolutely no sense, the interface_mode is an attribute supported by the device. If you do not expose it to z2m, then you can't set it. If you can't set it, then how do you tell the meter if it is gas or electricity? e.g. the probe used.


> .withValueMin(0).withValueMax(65535) - sliders are not working correctly in zigbee2Mqtt. It's sending 2-3 requests, and only first one contains selected value.

They were a slider just for testing purposes, once it is confirmed working this will be changed to an enum, allowing users to only use expected values.


> msg.data.hasOwnProperty - is not populating selected value, because there are no property in message.

Is this with the new code that includes the default endpoint? Cause if this is with the old code without the endpoint, then yeah, that is correct.



For people using this for electricity only, then nothing else is needed apart from changing the pulse_configuration, but the device is capable of more, hence our work here :)

@lanky8804 Thanks for confirming. Can you please keep an eye on the logs, and tell me, when you press the refresh button next to pulse_configuration, and interface_mode, what are you getting? Are you seeing values or errors. From here I can improve on the code and then help you have official support. Thanks again

**Comment by robertalexa** [comment]
> They were a slider just for testing purposes, once it is confirmed working this will be changed to an enum, allowing users to only use expected values.

By this I am refering to the interface_mode. The pulse_configuration will remain as is. Sliders should not be used like that, you should type in the value in the box instead of dragging, especially when the sliders supports thousands of values. Or alternatively, and maybe the best way, is to send the payload via mqtt as a singular controlled message.

**Comment by lanky8804** [comment]
@robertalexa been changing the pulse rate and refreshing, no errors occuring atm from clicking refresh next to pulse_configuration.

Some Logs:
`o MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"Pulse Configuration":3200,"develcoPulseConfiguration":3200,"energy":1348.25,"linkquality":0,"power":682}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"Pulse Configuration":3200,"develcoPulseConfiguration":3200,"energy":1348.25,"linkquality":0,"power":679}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"Pulse Configuration":3200,"develcoPulseConfiguration":3200,"energy":1348.25,"linkquality":0,"power":664}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"Pulse Configuration":3200,"develcoPulseConfiguration":3200,"energy":1348.25,"linkquality":0,"power":653}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"Pulse Configuration":3200,"develcoPulseConfiguration":3200,"energy":1348.25,"linkquality":0,"power":644}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"Pulse Configuration":3200,"develcoPulseConfiguration":5000,"energy":1348.25,"linkquality":0,"power":644}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"Pulse Configuration":5000,"develcoPulseConfiguration":5000,"energy":1348.25,"linkquality":0,"power":644}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"Pulse Configuration":5000,"develcoPulseConfiguration":5000,"energy":862.88,"linkquality":0,"power":406}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"Pulse Configuration":5000,"develcoPulseConfiguration":5000,"energy":862.88,"linkquality":0,"power":407}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"Pulse Configuration":5000,"develcoPulseConfiguration":5000,"energy":862.88,"linkquality":0,"power":407}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"Pulse Configuration":5000,"develcoPulseConfiguration":5000,"energy":862.88,"linkquality":0,"power":408}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"Pulse Configuration":5000,"develcoPulseConfiguration":3200,"energy":862.88,"linkquality":0,"power":408}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"Pulse Configuration":3200,"develcoPulseConfiguration":3200,"energy":862.88,"linkquality":0,"power":408}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"Pulse Configuration":3200,"develcoPulseConfiguration":3200,"energy":1348.26,"linkquality":0,"power":635}'`

**Comment by zuzuman** [comment]
> @zuzuman Have you actually tried the code or are you just giving me your thoughts?
Yes. I debugged 4 hours my sensor, and now - it's storing some rand dummy data (and publishing it) 
![image](https://user-images.githubusercontent.com/3532812/141113473-b79745f2-a460-4157-9af5-8c343bf04b6f.png)


> This makes absolutely no sense, the interface_mode is an attribute supported by the device. If you do not expose it to z2m, then you can't set it. If you can't set it, then how do you tell the meter if it is gas or electricity? e.g. the probe used.
Yes, but it's now used in zigbee2mqtt. It's used by Deconz or develco original app - for knowing, what value to show (m3 or kWh)



> For people using this for electricity only, then nothing else is needed apart from changing the pulse_configuration, but the device is capable of more, hence our work here :)
Only two values changing something - summary metering (reseting counter), and pulse configuration (changing impulse counting)

**Comment by robertalexa** [comment]
@lanky8804 Thanks for the logs provided. Can you confirm everything is ok for interface_mode too?

@zuzuman I have a feeling that you might not understand this enough, because you are sending contradictory messages.


> Yes. I debugged 4 hours my sensor, and now - it's storing some rand dummy data (and publishing it)

I posted this 2 hours ago, so how did you debug for 4? The code from this morning is EDITED, it is not the same. The dummy data is not dummy, is data saved agaisnt the device in the database file from previous converters. Remove the device, restart z2m, pair the device, old data is gone



> Yes, but it's not used in zigbee2mqtt. It's used by Deconz or develco original app - for knowing, what value to show (m3 or kWh)

Exactly, it needs to be added to z2m. This is exactly what we are doing now, getting this device and ALL its functions supported by z2m. You personally probably only care about electricity, but the device supports more, thats why it needs to be added. So another person buying the device and wanting to measure gas or water, can do it :)



> Only two values changing something - summary metering (reseting counter), and pulse configuration (changing impulse counting)

Use the right terms of the attributes, to not cause confusion:
pulse_configuration - number of pulses
interface_mode - set what probe is used to measure what. e.g. elec, gas, water etc
current_summation - if I understand this correctly is so that you set how much the devices counted so far. This number gets shows on the display (from what i understand, i do not have the device), and it keeps growing as you use more. This will only reset the counter is the value you add is wrong.

These 3 attributes go hand in hand, they all rely on each other. If you still have doubts, I will do my best to help :)

**Comment by robertalexa** [comment]
I think I now have a final version of the converter:

```javascript
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;

const constants_develcoInterfaceMode = {
    0: 'electricity',
    1: 'gas',
    2: 'water',
    256: 'kamstrup-kmp',
    257: 'linky',
    258: 'IEC62056-21',
    259: 'DSMR-2.3',
    260: 'DSMR-4.0'
};

const fzLocal = {
    develco_metering: {
        cluster: 'seMetering',
        type: ['attributeReport', 'readResponse'],
        convert: (model, msg, publish, options, meta) => {
            const result = {};
            if (msg.data.hasOwnProperty('develcoPulseConfiguration')) {
                result[postfixWithEndpointName('pulse_configuration', msg, model)] =
                    msg.data['develcoPulseConfiguration'];
            }
            if (msg.data.hasOwnProperty('develcoInterfaceMode')) {
                result[postfixWithEndpointName('interface_mode', msg, model)] =
                    msg.data['develcoInterfaceMode'];
            }
            return result;
        },
    },
}

const tzLocal = {
    develco_pulse_configuration: {
        key: ['pulse_configuration'],
        convertSet: async (entity, key, value, meta) => {
            await entity.write('seMetering', {'develcoPulseConfiguration': value}, {manufacturerCode: 0x1015});
            return {readAfterWriteTime: 200, state: {'pulse_configuration': value}};
        },
        convertGet: async (entity, key, meta) => {
            await entity.read('seMetering', ['develcoPulseConfiguration'], {manufacturerCode: 0x1015});
        },
    },
    develco_interface_mode: {
        key: ['interface_mode'],
        convertSet: async (entity, key, value, meta) => {
            const payload = {'develcoInterfaceMode': utils.getKey(constants_develcoInterfaceMode, value, undefined, Number)};
            await entity.write('seMetering', payload, {manufacturerCode: 0x1015});
            return {readAfterWriteTime: 200, state: {'interface_mode': value}};
        },
        convertGet: async (entity, key, meta) => {
            await entity.read('seMetering', ['develcoInterfaceMode'], {manufacturerCode: 0x1015});
        },
    },
    develco_current_summation: {
        key: ['current_summation'],
        convertSet: async (entity, key, value, meta) => {
            await entity.write('seMetering', {'develcoCurrentSummation': value}, {manufacturerCode: 0x1015});
            return {state: {'current_summation': value}};
        }
    },
}

const definition = {
    zigbeeModel: ['ZHEMI101'],
    model: 'ZHEMI101',
    vendor: 'Develco',
    description: 'Energy meter',
    fromZigbee: [fz.metering, fzLocal.develco_metering],
    toZigbee: [tzLocal.develco_pulse_configuration, tzLocal.develco_interface_mode],
    endpoint: (device) => {
        return {'default': 2};
    },
    configure: async (device, coordinatorEndpoint, logger) => {
        const endpoint = device.getEndpoint(2);
        await reporting.bind(endpoint, coordinatorEndpoint, ['seMetering']);
        await reporting.instantaneousDemand(endpoint);
        await reporting.readMeteringMultiplierDivisor(endpoint);
    },
    exposes: [
        e.power(), 
        e.energy(),
        exposes.numeric('pulse_configuration', ea.ALL).withValueMin(0).withValueMax(65535)
            .withDescription('Pulses per kwh. Default 1000 imp/kWh'),
        exposes.enum('interface_mode', ea.ALL,
                ['electricity', 'gas', 'water', 'kamstrup-kmp', 'linky', 'IEC62056-21', 'DSMR-2.3', 'DSMR-4.0'])
                .withDescription('Probe used/what is being measured'),
        exposes.numeric('current_summation', ea.SET)
    ],
};

module.exports = definition;
```

I would appreciate if you guys could test it.

**Comment by robertalexa** [comment]
Sorry that would throw an error. Updated the code now.

**Comment by lanky8804** [comment]
@robertalexa Just updated code again, still looks all good.

Could you please confirm what/where interface_mode is and I will test.

**Comment by robertalexa** [comment]
It should be available in z2m Frontend. That should make it easy to check.

https://www.zigbee2mqtt.io/guide/configuration/frontend.html

**Comment by lanky8804** [comment]
@robertalexa looks to be all ok, this is what i have.

![image](https://user-images.githubusercontent.com/4680961/141129876-037bbf93-fc05-4577-8b78-64c6f5a92842.png)

**Comment by robertalexa** [comment]
Can you click on the energy monitor and try to set values for pulse and interface_mode. Dont forget to revert interface mode to what you currently have. Also click the refresh icons and see what happens.

Thanks a lot

**Comment by lanky8804** [comment]
@robertalexa sorry for the delay. So getting some errors when changing interface_mode. See log below.

`o MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":1362.15,"interface_mode":null,"linkquality":0,"power":462,"pulse_configuration":null}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":1362.16,"interface_mode":null,"linkquality":0,"power":463,"pulse_configuration":null}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":1362.16,"interface_mode":null,"linkquality":0,"power":463,"pulse_configuration":3200}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":1362.16,"interface_mode":null,"linkquality":0,"power":462,"pulse_configuration":3200}'
Error Publish 'set' 'interface_mode' to 'Frient - Energy Monitor' failed: 'Error: Unknown attribute 'develcoInterfaceMode', specify either an existing attribute or a number'
Info MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"message":"Publish 'set' 'interface_mode' to 'Frient - Energy Monitor' failed: 'Error: Unknown attribute 'develcoInterfaceMode', specify either an existing attribute or a number'","meta":{"friendly_name":"Frient - Energy Monitor"},"type":"zigbee_publish_error"}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":1362.16,"interface_mode":null,"linkquality":0,"power":462,"pulse_configuration":3200}'
Error Publish 'set' 'interface_mode' to 'Frient - Energy Monitor' failed: 'Error: Unknown attribute 'develcoInterfaceMode', specify either an existing attribute or a number'
Info MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"message":"Publish 'set' 'interface_mode' to 'Frient - Energy Monitor' failed: 'Error: Unknown attribute 'develcoInterfaceMode', specify either an existing attribute or a number'","meta":{"friendly_name":"Frient - Energy Monitor"},"type":"zigbee_publish_error"}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":1362.16,"interface_mode":null,"linkquality":0,"power":463,"pulse_configuration":3200}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":1362.16,"interface_mode":null,"linkquality":0,"power":464,"pulse_configuration":3200}'
Error Publish 'get' 'interface_mode' to 'Frient - Energy Monitor' failed: 'Error: Cluster 'seMetering' has no attribute 'develcoInterfaceMode''
Info MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"message":"Publish 'get' 'interface_mode' to 'Frient - Energy Monitor' failed: 'Error: Cluster 'seMetering' has no attribute 'develcoInterfaceMode''","meta":{"friendly_name":"Frient - Energy Monitor"},"type":"zigbee_publish_error"}'`

**Comment by lanky8804** [comment]
The power reading seems more accurate now, not sure if you changed anything?

My house idles at 400-500w.

Previous code said was arrounf 650w, nothing extra was on so house was idle.

The latest code is back to around 460w which is what i expect.

**Comment by robertalexa** [comment]
@lanky8804

1. I have updated the code above, it had a small error, but unrelated to your error.
2. What version of zigbee2mqtt are u on? I suspect you are on an older one based on `Error: Unknown attribute 'develcoInterfaceMode'` This attribute should be available in the latest version of z2m
3. Nothing about my code would have changed the power value. Only the values you are setting (like the interface mode or the pulse configuration) would impact the operation of the device, and thus the readings it provides

**Comment by lanky8804** [comment]
1. Using latest code above.
2. Using Zigbee Edge - Zigbee2MQTT version 1.21.1-dev commit: bb5be4d
3. Ok odd, will monitor

**Comment by robertalexa** [comment]
@lanky8804 

Can you please update z2m to latest version? In your version zigbee-herdsman is at version 0.13.144
![image](https://user-images.githubusercontent.com/9019306/141289665-f6c17652-cde9-4496-ae82-371d4ad6ffed.png)

And I have added interface mode as of version 0.13.145
![image](https://user-images.githubusercontent.com/9019306/141289744-a8b375e2-2506-43ba-91b5-129b72191170.png)

Simply updating zigbee2mqtt to latest version will get you the correct software.

After that, try changing the interface mode and report your findings :)

Thanks

**Comment by lanky8804** [comment]
Updated, getting this error however the setting still applies and doesnt vanish.

`Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":1362.92,"interface_mode":"electricity","linkquality":0,"power":513,"pulse_configuration":3200}'
Error Exception while calling fromZigbee converter: postfixWithEndpointName is not defined}
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":1362.92,"interface_mode":"electricity","linkquality":0,"power":512,"pulse_configuration":3200}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":1362.92,"interface_mode":"electricity","linkquality":0,"power":512,"pulse_configuration":3200}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":1362.92,"interface_mode":"electricity","linkquality":0,"power":511,"pulse_configuration":3200}'`

**Comment by robertalexa** [comment]
@lanky8804 based on that error, i would expect you to be getting a similar thing when you change `pulse_configuration`. Can you please confirm that?

Also, while i have a thinker at my end, can you make sure that while testing you have the log mode set to `debug` rather than `info`

And also, when pasting multiline, instead of using the single tick, use the 3 tick. Here is a screenshot to understand:
![image](https://user-images.githubusercontent.com/9019306/141302625-4c01fe57-d030-49b1-8440-4bb56bb91a95.png)

Thanks for your help with this :)

**Comment by robertalexa** [comment]
Update code here:
```javascript
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const utils = require('zigbee-herdsman-converters/lib/utils');
const e = exposes.presets;
const ea = exposes.access;

const constants_develcoInterfaceMode = {
    0: 'electricity',
    1: 'gas',
    2: 'water',
    256: 'kamstrup-kmp',
    257: 'linky',
    258: 'IEC62056-21',
    259: 'DSMR-2.3',
    260: 'DSMR-4.0'
};

const fzLocal = {
    develco_metering: {
        cluster: 'seMetering',
        type: ['attributeReport', 'readResponse'],
        convert: (model, msg, publish, options, meta) => {
            const result = {};
            if (msg.data.hasOwnProperty('develcoPulseConfiguration')) {
                result[utils.postfixWithEndpointName('pulse_configuration', msg, model)] =
                    msg.data['develcoPulseConfiguration'];
            }
            if (msg.data.hasOwnProperty('develcoInterfaceMode')) {
                result[utils.postfixWithEndpointName('interface_mode', msg, model)] =
                    msg.data['develcoInterfaceMode'];
            }
            return result;
        },
    },
}

const tzLocal = {
    develco_pulse_configuration: {
        key: ['pulse_configuration'],
        convertSet: async (entity, key, value, meta) => {
            await entity.write('seMetering', {'develcoPulseConfiguration': value}, {manufacturerCode: 0x1015});
            return {readAfterWriteTime: 200, state: {'pulse_configuration': value}};
        },
        convertGet: async (entity, key, meta) => {
            await entity.read('seMetering', ['develcoPulseConfiguration'], {manufacturerCode: 0x1015});
        },
    },
    develco_interface_mode: {
        key: ['interface_mode'],
        convertSet: async (entity, key, value, meta) => {
            const payload = {'develcoInterfaceMode': utils.getKey(constants_develcoInterfaceMode, value, undefined, Number)};
            await entity.write('seMetering', payload, {manufacturerCode: 0x1015});
            return {readAfterWriteTime: 200, state: {'interface_mode': value}};
        },
        convertGet: async (entity, key, meta) => {
            await entity.read('seMetering', ['develcoInterfaceMode'], {manufacturerCode: 0x1015});
        },
    },
    develco_current_summation: {
        key: ['current_summation'],
        convertSet: async (entity, key, value, meta) => {
            await entity.write('seMetering', {'develcoCurrentSummation': value}, {manufacturerCode: 0x1015});
            return {state: {'current_summation': value}};
        }
    },
}

const definition = {
    zigbeeModel: ['ZHEMI101'],
    model: 'ZHEMI101',
    vendor: 'Develco',
    description: 'Energy meter',
    fromZigbee: [fz.metering, fzLocal.develco_metering],
    toZigbee: [tzLocal.develco_pulse_configuration, tzLocal.develco_interface_mode, tzLocal.develco_current_summation],
    endpoint: (device) => {
        return {'default': 2};
    },
    configure: async (device, coordinatorEndpoint, logger) => {
        const endpoint = device.getEndpoint(2);
        await reporting.bind(endpoint, coordinatorEndpoint, ['seMetering']);
        await reporting.instantaneousDemand(endpoint);
        await reporting.readMeteringMultiplierDivisor(endpoint);
    },
    exposes: [
        e.power(), 
        e.energy(),
        exposes.numeric('pulse_configuration', ea.ALL).withValueMin(0).withValueMax(65535)
            .withDescription('Pulses per kwh. Default 1000 imp/kWh'),
        exposes.enum('interface_mode', ea.ALL,
                ['electricity', 'gas', 'water', 'kamstrup-kmp', 'linky', 'IEC62056-21', 'DSMR-2.3', 'DSMR-4.0'])
                .withDescription('Probe used/what is being measured'),
        exposes.numeric('current_summation', ea.SET)
    ],
};

module.exports = definition;
```

**Comment by lanky8804** [comment]
Have just updated to new code and changed log to debug.

Changing the pulse_configuration and interface_mode doesnt appear to error anymore.

One thing i have noticed is that if i refresh the interface_mode then it wipes the value selected e.g. elec, gas. But is seleced from drop down then saves it.

```
Debug Received MQTT message on 'zigbee2mqtt/Frient - Energy Monitor/set' with data '{"interface_mode":"electricity"}'
Debug Publishing 'set' 'interface_mode' to 'Frient - Energy Monitor'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":1364.62,"interface_mode":"electricity","linkquality":0,"power":674,"pulse_configuration":3200}'
Debug Received Zigbee message from 'Frient - Energy Monitor', type 'attributeReport', cluster 'seMetering', data '{"currentSummDelivered":[0,1364616],"currentTier1SummDelivered":[65535,4294967295],"currentTier2SummDelivered":[65535,4294967295],"instantaneousDemand":673,"status":0}' from endpoint 2 with groupID 0
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":1364.62,"interface_mode":"electricity","linkquality":0,"power":673,"pulse_configuration":3200}'
Debug Received MQTT message on 'zigbee2mqtt/Frient - Energy Monitor/get' with data '{"interface_mode":""}'
Debug Publishing get 'get' 'interface_mode' to 'Frient - Energy Monitor'
Debug Received Zigbee message from 'Frient - Energy Monitor', type 'readResponse', cluster 'seMetering', data '{"develcoInterfaceMode":0}' from endpoint 2 with groupID 0
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":1364.62,"interface_mode":0,"linkquality":0,"power":673,"pulse_configuration":3200}'
Debug Received Zigbee message from 'Frient - Energy Monitor', type 'attributeReport', cluster 'seMetering', data '{"currentSummDelivered":[0,1364617],"currentTier1SummDelivered":[65535,4294967295],"currentTier2SummDelivered":[65535,4294967295],"instantaneousDemand":673,"status":0}' from endpoint 2 with groupID 0
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":1364.62,"interface_mode":0,"linkquality":0,"power":673,"pulse_configuration":3200}'
Debug Received Zigbee message from 'Frient - Energy Monitor', type 'attributeReport', cluster 'seMetering', data '{"currentSummDelivered":[0,1364618],"currentTier1SummDelivered":[65535,4294967295],"currentTier2SummDelivered":[65535,4294967295],"instantaneousDemand":674,"status":0}' from endpoint 2 with groupID 0
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":1364.62,"interface_mode":0,"linkquality":0,"power":674,"pulse_configuration":3200}'
Debug Received MQTT message on 'zigbee2mqtt/Frient - Energy Monitor/set' with data '{"interface_mode":"electricity"}'
Debug Publishing 'set' 'interface_mode' to 'Frient - Energy Monitor'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":1364.62,"interface_mode":"electricity","linkquality":0,"power":674,"pulse_configuration":3200}'
Debug Received Zigbee message from 'Frient - Energy Monitor', type 'attributeReport', cluster 'seMetering', data '{"currentSummDelivered":[0,1364620],"currentTier1SummDelivered":[65535,4294967295],"currentTier2SummDelivered":[65535,4294967295],"instantaneousDemand":674,"status":0}' from endpoint 2 with groupID 0
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":1364.62,"interface_mode":"electricity","linkquality":0,"power":674,"pulse_configuration":3200}'
```

**Comment by robertalexa** [comment]
@lanky8804 Thank you for testing and for the debug log. It helped me understand the issue that you are describing about losing the value.

I now have updated code that should fix it. I hopefully this will be the final working version. If it is, please confirm and I will make sure to submit a PR and get your device officially improved as of the next version of z2m (1st of December)

```javascript
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const utils = require('zigbee-herdsman-converters/lib/utils');
const e = exposes.presets;
const ea = exposes.access;

const constants_develcoInterfaceMode = {
    0: 'electricity',
    1: 'gas',
    2: 'water',
    256: 'kamstrup-kmp',
    257: 'linky',
    258: 'IEC62056-21',
    259: 'DSMR-2.3',
    260: 'DSMR-4.0'
};

const fzLocal = {
    develco_metering: {
        cluster: 'seMetering',
        type: ['attributeReport', 'readResponse'],
        convert: (model, msg, publish, options, meta) => {
            const result = {};
            if (msg.data.hasOwnProperty('develcoPulseConfiguration')) {
                result[utils.postfixWithEndpointName('pulse_configuration', msg, model)] =
                    msg.data['develcoPulseConfiguration'];
            }

            if (msg.data.hasOwnProperty('develcoInterfaceMode')) {
                result[utils.postfixWithEndpointName('interface_mode', msg, model)] =
                    constants_develcoInterfaceMode.hasOwnProperty(msg.data['develcoInterfaceMode']) ?
                        constants_develcoInterfaceMode[msg.data['develcoInterfaceMode']] :
                        msg.data['develcoInterfaceMode'];
            }

            return result;
        },
    },
}

const tzLocal = {
    develco_pulse_configuration: {
        key: ['pulse_configuration'],
        convertSet: async (entity, key, value, meta) => {
            await entity.write('seMetering', {'develcoPulseConfiguration': value}, {manufacturerCode: 0x1015});
            return {readAfterWriteTime: 200, state: {'pulse_configuration': value}};
        },
        convertGet: async (entity, key, meta) => {
            await entity.read('seMetering', ['develcoPulseConfiguration'], {manufacturerCode: 0x1015});
        },
    },
    develco_interface_mode: {
        key: ['interface_mode'],
        convertSet: async (entity, key, value, meta) => {
            const payload = {'develcoInterfaceMode': utils.getKey(constants_develcoInterfaceMode, value, undefined, Number)};
            await entity.write('seMetering', payload, {manufacturerCode: 0x1015});
            return {readAfterWriteTime: 200, state: {'interface_mode': value}};
        },
        convertGet: async (entity, key, meta) => {
            await entity.read('seMetering', ['develcoInterfaceMode'], {manufacturerCode: 0x1015});
        },
    },
    develco_current_summation: {
        key: ['current_summation'],
        convertSet: async (entity, key, value, meta) => {
            await entity.write('seMetering', {'develcoCurrentSummation': value}, {manufacturerCode: 0x1015});
            return {state: {'current_summation': value}};
        }
    },
}

const definition = {
    zigbeeModel: ['ZHEMI101'],
    model: 'ZHEMI101',
    vendor: 'Develco',
    description: 'Energy meter',
    fromZigbee: [fz.metering, fzLocal.develco_metering],
    toZigbee: [tzLocal.develco_pulse_configuration, tzLocal.develco_interface_mode, tzLocal.develco_current_summation],
    endpoint: (device) => {
        return {'default': 2};
    },
    configure: async (device, coordinatorEndpoint, logger) => {
        const endpoint = device.getEndpoint(2);
        await reporting.bind(endpoint, coordinatorEndpoint, ['seMetering']);
        await reporting.instantaneousDemand(endpoint);
        await reporting.readMeteringMultiplierDivisor(endpoint);
    },
    exposes: [
        e.power(), 
        e.energy(),
        exposes.numeric('pulse_configuration', ea.ALL).withValueMin(0).withValueMax(65535)
            .withDescription('Pulses per kwh. Default 1000 imp/kWh'),
        exposes.enum('interface_mode', ea.ALL,
                ['electricity', 'gas', 'water', 'kamstrup-kmp', 'linky', 'IEC62056-21', 'DSMR-2.3', 'DSMR-4.0'])
                .withDescription('Probe used/what is being measured'),
        exposes.numeric('current_summation', ea.SET)
    ],
};

module.exports = definition;
```

**Comment by lanky8804** [comment]
Ok so new code no errors appearing in logs but this now with interface_mode.

So if it was on elec and i select from drop down gas. It will still display elec until i click the refresh button. Is this intended.

**Comment by robertalexa** [comment]
That is not intended no, but with the latest code it should not happen either. The value is meant to update as soon as the "writing" is complete on the device.

Could you please provide some logs? (After changing the dropdown give it some time, don't press refresh for a while)

I have also noticed in your logs that the linkquality is always 0, bad connection quality could be  the reason for the issue you are describing, but you would also struggle with other commands too, so very weird.

**Comment by lanky8804** [comment]
Ok yes after about 3-5s it updates. Log below, like i say no error.

Yes i dont think the LQ is correct or updating as it is about 3m away from the device with a direct line or sight and no obsticals in the way. The power and energy update every few seconds so would indicate that it has a good signal.

```
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":1370.62,"interface_mode":"electricity","linkquality":0,"power":230,"pulse_configuration":3200}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":1370.62,"interface_mode":"gas","linkquality":0,"power":230,"pulse_configuration":3200}'
Info MQTT publish: topic 'zigbee2mqtt/Frient - Energy Monitor', payload '{"energy":1370.62,"interface_mode":"electricity","linkquality":0,"power":230,"pulse_configuration":3200}'
```

**Comment by robertalexa** [comment]
> Ok yes after about 3-5s it updates. Log below, like i say no error.

Happy to hear. I have created the PR for official support https://github.com/Koenkk/zigbee-herdsman-converters/pull/3337

When this will be supported natively with the next z2m update, you can safely remove the external converter, the device will just work. If you are happy with everything, could you please close this issue?

Thanks for your help with testing this. Hopefully the new features will benefit you

About LQ:

Looking at other people's logs https://github.com/Koenkk/zigbee2mqtt/issues/8638#issuecomment-920088522 it feels like the issue might be sth at your end. I would suggest that you un-pair the device, reboot the device, restart z2m, pair it again, and see if that fixes it.

**Comment by lanky8804** [comment]
@robertalexa Excellent thank you for you help with resolving this.
